Switching updates

S-S-X opened the pull request:
## NEEDS SOME TESTING

So I've not tested this much, just basics with very simple stuff.

Reduces metadata use. Removes completely following metadata keys:
* LV_EU_timeout
* MV_EU_timeout
* HV_EU_timeout
* LV_network
* MV_network
* HV_network

Overload reset clean up cable network.
~Overload causes early exit from network traversal.~ Reverted that, should be able to enable again now that cable cache is also cleaned up but better to not include that yet.

Fixes #76 
Partially fixes #77

S-S-X commented:
Early return from network traversal could cause unwanted side effects, removing network from `technic.cables` should be enough to mitigate this but is not actually very good way to do it.
Better approach would be to use network to cables mapping and only run through cables belonging to single network, however looping through all cables and comparing table values is probably still better than earlier mixed lua tables and per node metadata approach.

Network stuff should be moved away from switching_station.lua to some network.lua which should provide clean API for all network related stuff and caching. That stuff is still scattered around multiple files like switching_station.lua and cables.lua which makes maintenance and improvements harder.

S-S-X commented on technic/machines/switching_station.lua:
```diff
@@ -135,36 +141,38 @@ local function add_cable_node(nodes, pos, network_id, queue)
 			-- delete caches for conflicting network
 			technic.networks[net_id_old] = nil
-			meta:set_string(net_id_key, net_id_new)
+			technic.networks[net_id_new] = nil
```
```
technic.networks[net_id_old] = nil
technic.networks[net_id_new] = nil
```
These actually should not be cleared here, also if early return fires back `technic.networks[net_id_new] = nil` cannot be here as it is network that is currently processed.

Function `reset_overloaded(network_id)` already clears these after restart timer has expired for network in question.
Both could still be used to clean up cables in a more controlled way (`technic.networks[network_id].all_nodes`).

OgelGames commented:
Still has the overload loop bug, except now it sometimes gets out of the loop, and I can't reproduce it consistently 😕

S-S-X commented:
> Still has the overload loop bug, except now it sometimes gets out of the loop, and I can't reproduce it consistently

Pretty sure this happens because cleanup is not done well enough, basically what I added to last 2 comments here...
If this seems to be problem then it could be option to leave early return out and let it check whole network even after conflict is detected.

Maybe I could still try to improve network cleanup a bit based on what I said in last source comment but not sure how well it goes...

BuckarooBanzay commented:
I switched the test server to the `switching` branch of the technic mod and played around a bit.
I hit the overload issue a few times with my existing circuits, looks like it does not like different cabling around another tier:

![screenshot_20200905_192904](https://user-images.githubusercontent.com/39065740/92310411-40e02200-efae-11ea-9670-7bc6ef314879.png)

It only overloads if the mv-cable and second switching station is placed above, but i think that has to do with the multi switching station issue you mentioned...

S-S-X commented:
> It only overloads if the mv-cable and second switching station is placed above, but i think that has to do with the multi switching station issue you mentioned...

Kind of yes but overload is basically caused because switching station belongs to all tiers, it is special case that I didn't handle when I removed metadata stuff:

network id was stored like this, note `tier` variable which also marked actual switching station tier:
```
local net_id_old = meta:get_string(tier.."_network")
```
and when I removed metadata it turned into this, note that this does not make difference between tiers:
```
local net_id_old = technic.cables[minetest.hash_node_position(pos)]
```

Simplest workaround would be to simply skip switching station, supply converter and power monitor.... does not sound nice, I'll see if I can find way that is more than just workaround.
Best would probably be to actually check for cable connections, that would probably also fix (or at least help with fixing) battery bugs / solar array bugs and similar stuff but yes readily available switching station data would also fix this.

S-S-X commented on technic/machines/switching_station.lua:
```diff
@@ -202,14 +209,48 @@ local traverse_network = function(PR_nodes, RE_nodes, BA_nodes, SP_nodes, all_no
 end
 
-local touch_nodes = function(list, tier)
+function technic.pos2network(pos)
+	return pos and technic.cables[minetest.hash_node_position(pos)]
+end
+
+function technic.network2pos(network_id)
+	return network_id and minetest.get_position_from_hash(network_id)
+end
+
+function technic.network2sw_pos(network_id)
+	-- Return switching station position for network.
+	-- It is not guaranteed that position actually contains switching station.
+	local sw_pos = minetest.get_position_from_hash(network_id)
+	sw_pos.y = sw_pos.y + 1
+	return sw_pos
+end
```
These could be better defined in some common helper/utility file instead or even bettersome network.lua but there's no such thing, at least not yet.

S-S-X commented:
Not very serious test but wanted to see what it gives. Collected 830 samples, measured execution of `technic.switching_station_run` with not so well thought HV network containing some thousands cables, 2 switching stations, few hundred battery boxes, few hundred furnaces and bit over 300 solar arrays.
During tests lag penalties and polyfuse disabled.

Switching branch average time 14281 (219f772e1d19692d45d7458c31823454f4228f4d)
Master branch switching/overload average time 15261 (6b3e19e734accccc3165e02a9ebf085c01ba2dc9)

Same with heavy CPU throttling using `cpulimit -m -l 1` (limit cpu usage to 1%, engine skipped a lot of ABMs and globalstep didn't execute too often so only bit over 200 samples here)
Switching branch average time 77583 (219f772e1d19692d45d7458c31823454f4228f4d)
Master branch switching/overload average time 92503 (6b3e19e734accccc3165e02a9ebf085c01ba2dc9)

Seems that performance at least is not worse, however actual performance impact might come from cache building if any.

Still things to do: fix multi tier machines.

S-S-X commented:
Should be better now with multi tier machines but again, not tested everything.
* Power monitors are no longer part of actual power network but instead standalone machines.
* Supply converters bypass overload checks completely (_because there's no real multi tier machine support after metadata not used_).
* Switching station wont attempt to activate on_construct if another switch can be found from loaded area (_yeah, not best check but better than nothing when you place a lot of switching stations to same network... for whatever reason..._).
* Switching station connection is checked and it only considers connection from below, fixes: https://github.com/mt-mods/technic/pull/80#issuecomment-687639730
  > It only overloads if the mv-cable and second switching station is placed above, but i think that has to do with the multi switching station issue you mentioned...

@BuckarooBanzay would it be possible to update this to pandorabox test?

BuckarooBanzay commented:
Wow, that was fast :astonished: 

> Should be better now with multi tier machines but again, not tested everything.
> @BuckarooBanzay would it be possible to update this to pandorabox test?

I updated the test-server (the world is still on an old state but it should do for now) to the latest `switching` branch.
Judging from my tests and all the existing power networks i have it looks pretty good :+1:

Awesome work!
